stout/flags.cc

```diff
@@ -24,29 +23,80 @@ void Parser::AddAllOrExit(google::protobuf::Message* message) {
   for (int i = 0; i < descriptor->field_count(); i++) {
     const auto* field = descriptor->field(i);
 
-    const auto& flag = field->options().GetExtension(stout::v1::flag);
+    // We need this descriptor for the subcommand's logic.
+    const google::protobuf::OneofDescriptor* real_oneof_field =
+        field->real_containing_oneof();
+
+    // Check if a message given by someone contains oneof field which
+    // has the `subcommand` name.
+    if (real_oneof_field != nullptr
+        && real_oneof_field->name() == "subcommand") {
+      // Run through all fields of oneof and make sure that all of them
+      // have `stout.v1.subcommand` extension.
+      for (int i = 0; i < real_oneof_field->field_count(); ++i) {
+        const google::protobuf::FieldDescriptor* oneof_fd =
+            real_oneof_field->field(i);
+        if (!oneof_fd->options().HasExtension(stout::v1::subcommand)) {
+          std::cerr << "Oneof 'subcommand' must contain fields with"
+                    << " the following extension: (stout.v1.subcommand)."
+                    << " Another extensions are illegal."
+                    << std::endl;
+          exit(1);
+        } else {
+          // Check for missing 'names' and 'help'
+          // in (stout.v1.subcommand) extension.
+          const auto& subcommand =
+              oneof_fd->options().GetExtension(stout::v1::subcommand);
+
+          if (subcommand.names().empty()) {
+            std::cerr
+                << "Missing at least one flag name in 'names' for field '"
+                << oneof_fd->full_name() << "'"
+                << std::endl;
+            std::exit(1);
+          }
 
-    if (flag.names().empty()) {
-      std::cerr
-          << "Missing at least one flag name in 'names' for field '"
-          << field->full_name() << "'"
-          << std::endl;
-      std::exit(1);
-    }
+          if (subcommand.help().empty()) {
+            std::cerr
+                << "Missing flag 'help' for field '"
+                << oneof_fd->full_name() << "'"
+                << std::endl;
+            std::exit(1);
+          }
+        }
+      }
+    } else {
+      if (field->options().HasExtension(stout::v1::subcommand)) {
```

Comment:
This is great, we also want to check that there is no `stout::v1::flag` extension for subcommands!